File: C++/src/gilded_rose.cpp
#include "GildedRose/gilded_rose.h"

#include <memory>
#include <string>

#include "aged_brie_item.h"
#include "backstage_passes_item.h"
#include "normal_item.h"

namespace
{
const char *const kItemNormal = "Normal Item";
const char *const kItemAgedBrie = "Aged Brie";
const char *const kItemBackstagePasses = "Backstage passes to a TAFKAL80ETC concert";
const char *const kItemSulfuras = "Sulfuras, Hand of Ragnaros";
}  // namespace

GildedRose::GildedRose(const std::string &name, int days_remaining, int quality)
    : _name(name)
    , _days_remaining(days_remaining)
    , _quality(quality)
{
}

void GildedRose::tick()
{
    if (_name == kItemNormal) {
        std::unique_ptr<Item> item{std::make_unique<NormalItem>(_days_remaining, _quality)};
        item->tick();
        _days_remaining = item->days_remaining();
        _quality = item->quality();
        return;
    }

    if (_name == kItemAgedBrie) {
        std::unique_ptr<Item> item{std::make_unique<AgedBrieItem>(_days_remaining, _quality)};
        item->tick();
        _days_remaining = item->days_remaining();
        _quality = item->quality();
        return;
    }

    if (_name == kItemSulfuras) {
        return;
    }

    if (_name == kItemBackstagePasses) {
        std::unique_ptr<Item> item{std::make_unique<BackstagePassesItem>(_days_remaining, _quality)};
        item->tick();
        _days_remaining = item->days_remaining();
        _quality = item->quality();
        return;
    }
}

int GildedRose::days_remaining()
{
    return _days_remaining;
}

int GildedRose::quality()
{
    return _quality;
}

File: Ruby/lib/gilded_rose.rb
require_relative("item")
require_relative("normal_item")
require_relative("aged_brie")
require_relative("backstage")
require_relative("sulfuras")
require_relative("conjured_mana")

class GildedRose
  ITEM_CLASSES = {
    "Aged Brie"                                 => AgedBrie,
    "Sulfuras, Hand of Ragnaros"                => Sulfuras,
    "Backstage passes to a TAFKAL80ETC concert" => Backstage,
    "Conjured Mana Cake"                        => ConjuredMana
  }

  def initialize(name:, days_remaining:, quality:)
    item = Item.new(name, days_remaining, quality)
    @current_item = class_from(name: name).new(item)
  end

  def tick
    @current_item.update
  end

  def days_remaining
    @current_item.item.days_remaining
  end

  def quality
    @current_item.item.quality
  end

  private

  def class_from(name:)
    ITEM_CLASSES[name] || NormalItem
  end
end

gilded_rose = GildedRose.new(name: "Conjured Mana Cake", days_remaining: 5, quality: 10)
gilded_rose.tick
